### ISSUE TITLE
A 'top-level' scope in a macro in item position unintuitively ignores everything

### ISSUE THREAD

**Issue by huonw** [rust-lang/rust#17022]
``` rust
#![feature(macro_rules)]

macro_rules! x {
    () => {{
        struct Foo;
    }}
}

x!()
```

```
<anon>:4:12: 4:13 error: macro expansion ignores token `{` and any following
<anon>:4     () => {{
                    ^
```

i.e. the parser is seeing the `{` and bailing on parsing an item, and the macro expander is diligently reporting that those tokens were ignored. It would be better if we could emit a more helpful message for this case specifically, since expression macros require the `{`s to be able to contain multiple things so users used to this can get tripped up (item macros don't need scopes for multiple items, and doing this would restrict those items to that inner scope anyway).

**Comment by iopq** [comment]
This is the top search result on this issue, but I'm a little confused on how to fix it. Can you suggest a work-around until it's fixed?

**Comment by durka** [comment]
Triage
===

For @iopq (hope you have moved on by now 😝 ) and others, the fix is to remove the extra set of braces, assuming the macro body consists of items (if it is statements or expressions, you can't use this macro outside of a function anyway).

New code: (nearly the same)

```rust
macro_rules! x {
    () => {{
        struct Foo;
    }}
}

x!();
```

The same perplexing error is printed but also

```
<anon>:7:1: 7:6 note: caused by the macro expansion here; the usage of `x!` is likely invalid in item context
<anon>:7 x!();
```

which tells you what is wrong but gives no hint as to how to fix it. Ideally, the compiler would drop the weasel words and tell you _why_ the macro is illegal in item context, as well as a help saying that double braces aren't required when a macro generates solely items.

**Comment by huonw** [comment]
This looks to have been accidentally closed by a sub-tree update misinterpreting "fixes #17022" in a rust-analyzer commit.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=224f0e3ad5f0a2c7bb21b4be1ee45994

```
error: macro expansion ignores token `{` and any following
 --> src/lib.rs:2:12
  |
2 |     () => {{
  |            ^
...
7 | x!();
  | ---- caused by the macro expansion here
  |
  = note: the usage of `x!` is likely invalid in item context
```

This might be fine to stay closed, though 😄
